Makefile

```diff
@@ -1,32 +1,48 @@
 FILES = prelude prepare solution template test
+BDIR = bin
+SDIR = src
+TDIR = tests
+INPUT = input.ml
+
 SUFFIX = .native
 SOLPREFIX = sol_
 MAPPREFIX = mapper_
 ASTPREFIX = ptree_
+#TARGETS = $(addsuffix $(SUFFIX), $(addprefix $(BDIR)/$(MAPPREFIX), $(FILES)))
 TARGETS = $(addsuffix $(SUFFIX), $(addprefix $(MAPPREFIX), $(FILES)))
 
-INPUT = input.ml
 
-OCB_LIBS = -package compiler-libs.common
-OCB = ocamlbuild $(OCB_LIBS)
+NAME = autogen
+EXE = $(NAME)$(SUFFIX)
+BEXE = $(BDIR)/$(EXE)
 
-DIRS = $(wildcard tests/*/)
+OCB_LIBS = -package compiler-libs.common -package cmdliner
+OCB_DIR_FLAGS = -I $(SDIR)
+OCB = ocamlbuild $(OCB_LIBS) $(OCB_DIR_FLAGS)
+
+TESTS = $(wildcard $(TDIR)/*/)
 
 OCF = ocamlfind ppx_tools/rewriter
 
-.PHONY : clean clear all tests
+.PHONY : all clean clear test tests_clean
 
 all : $(TARGETS)
+	$(OCB) $(EXE)
+	ln -sf $(EXE) $(NAME)
```

Comment:
Here, I would add something like `mkdir -p $(BDIR) && mv $(EXE) $(BEXE)` to install the executable in the proper subdirectory.